biorxiv/journal_tracker/journal_recommendation_experiment.py

```diff
@@ -0,0 +1,354 @@
+
+from annorxiver_modules.journal_rec_helper import (
+    cross_validation, dummy_evaluate,
+    knn_evaluate, knn_centroid_evaluate
+)
+
+
+# # Load bioRxiv Document Vectors
+
+# In[2]:
+
+
+biorxiv_journal_df = (
+    pd.read_csv("output/mapped_published_doi.tsv", sep="\t")
+    .groupby("doi")
+    .agg({
+        "document":"last",
+        "category":"first",
+        "journal":"first",
+        "doi":"last",
+        "published_doi":"first",  
+        "pmcid":"first", 
+    })
+    .reset_index(drop=True)
+)
+biorxiv_journal_df.head()
+
+
+# In[3]:
+
+
+# Count number of Non-NaN elements
+print(f"Number of Non-NaN entries: {biorxiv_journal_df.pmcid.count()}")
+print(f"Total number of entries: {biorxiv_journal_df.shape[0]}")
+print(f"Percent Covered: {biorxiv_journal_df.pmcid.count()/biorxiv_journal_df.shape[0]:.2f}")
+
+
+# In[4]:
+
+
+golden_set_df = biorxiv_journal_df.query("pmcid.notnull()")
+golden_set_df.head()
+
+
+# # Load Pubmed Central Document Vectors
+
+# In[5]:
+
+
+pmc_articles_df = (
+    pd.read_csv(
+        "../../pmc/exploratory_data_analysis/output/pubmed_central_journal_paper_map.tsv.xz", 
+        sep="\t"
+    )
+    .query("article_type=='research-article'")
+)
+print(pmc_articles_df.shape)
+pmc_articles_df.head()
+
+
+# In[6]:
+
+
+journals = pmc_articles_df.journal.value_counts()
+print(journals.shape)
+journals
+
+
+# In[7]:
+
+
+# Filter out low count journals
+pmc_articles_df = pmc_articles_df.query(f"journal in {journals[journals > 100].index.tolist()}")
+print(pmc_articles_df.shape)
+pmc_articles_df.head()
+
+
+# In[8]:
+
+
+pmc_embedding_dict = {
+    int(path.stem[-7:-4]):pd.read_csv(
+        str(path), 
+        sep="\t"
+    )
+    for path in Path("../../pmc/word_vector_experiment/output/").rglob("*tsv.xz")
+}
+pmc_embedding_dict[300].head()
+
+
+# In[9]:
+
+
+full_training_dataset = {
+    dim : (
+        pmc_articles_df
+        .query(f"pmcid not in {golden_set_df.pmcid.tolist()}")
+        [["pmcid", "journal"]]
+        .merge(pmc_embedding_dict[dim], left_on="pmcid", right_on="document")
+        .drop("pmcid", axis=1)
+        .set_index("document")
+    )
+    for dim in pmc_embedding_dict
+}
+
+
+# In[10]:
+
+
+subsampled_training_dataset = {
+    dim : (
+        pmc_articles_df
+        .query(f"pmcid not in {golden_set_df.pmcid.tolist()}")
+        [["pmcid", "journal"]]
+        .merge(pmc_embedding_dict[dim], left_on="pmcid", right_on="document")
+        .drop("pmcid", axis=1)
+        .groupby("journal", group_keys=False)
+        .apply(lambda x: x.sample(min(len(x), 100), random_state=100))
+        .set_index("document")
+    )
+    for dim in pmc_embedding_dict
+}
+
+
+# # Train Recommendation System
+
+# In[11]:
+
+
+knn_model = KNeighborsClassifier(n_neighbors=10)
+
+
+# ## Random Journal Prediction
+
+# The central idea here is to answer the question what is the accuracy when journals are recommended at random?
+
+# In[12]:
+
+
+model = DummyClassifier(strategy='uniform')
```

Comment:
You're random model is assigning journal recommendations with a uniform distribution. i.e. article A is equally likely (1/N) to be assigned to any of the  N categories of journals

Reply:
>You're random model is assigning journal recommendations with a uniform distribution. i.e. article A is equally likely (1/N) to be assigned to any of the N categories of journals

Yup. You are correct. This is just to show what performance would be if all journals were equally likely to be selected.